lab-lindsay/lib/binarysearchtree.js

```diff
@@ -0,0 +1,181 @@
+'use strict';
+
+//creating a constructor for the tree node
+function Node(val) {
+  this.value = val
+  this.data = null
+  this.left = null
+  this.right = null
+  this.parent = null
+}
+
+//creating a constructor for the binary search tree
+function BinarySearchTree() {
+  this.root = null
+}
+
+//nodes to the left of the root are smaller values than the root
+//nodes to the right of the root are higher values than the root
+
+BinarySearchTree.prototype.push = function(val) {
+  let root = this.root
+  if(!root) {
+    this.root = new Node(val)
+    return;
+  }
+  let currentNode = root
+  let newNode = new Node(val)
+
+  while(currentNode) {
+    if(val < currentNode.value) {
+      if(!currentNode.left) {
+        currentNode.left = newNode
+        break
+      } else {
+        currentNode = currentNode.left
+      }
+    } else {
+      if(!currentNode.right) {
+        currentNode.right = newNode
+        break
+      } else {
+        currentNode = currentNode.right
+      }
+    }
+  }
+}
+
+//finding min node value in tree
+function minNode(node) {
+  if(!node) {
+    return 0
+  }
+  if(node.left) {
+    return minNode(node.left)
+  }
+  return node.value
+}
+
+//finding max node value in tree
+function maxNode(node) {
+  if(!node) {
+    return 0
+  }
+  if(node.right) {
+    return maxNode(node.right)
+  }
+  return node.value
+}
+
+//Binary tree from array method
+function bstFromArray(num) {
+  function recurse(num, start, end) {
+    if(start > end) {
+      return null;
+    }
+    let mid = Math.floor((start + end) / 2);
```

Comment:
This BST from array looks to imply that you're working with a sorted array, but may not be the case. Will it always be true that you want to work from the mid point of the array if you're unsure of sortedness? (is that a word?)